### PR TITLE
DRAFT: Feat/add new provider for remote generic versioned storage

### PR DESCRIPTION
**Pull request by andrewx82**
This is with reference to https://github.com/six7/figma-tokens/discussions/1112 . It exposes the same interface as the existing JSONBin storage provider, however it abstracts away the specifics on the JSONBin endpoint such as its hardcoded url , fixed headers ,etc and allows those to be set in the storage provider settings in the case that someone would like to create their own endpoint that is read/write. I understand the remote URL storage provider is ready only at the moment.  

@six7 One of my concerns is how to manage array data for the storage provider, your current  create and edit storageitem modals use a passed in onChange event which expects to be interacting with a raw input element, and then sets the property on the data object directly, which works for simple key value associations, but not particularly well when array or nested object operations occur. As such I've tested a prototype using a set operation notation which would be the least disruptive to the current approach, but I'm not happy as operations like deletion, etc are not easily integrated

Would you be willing to talk to me regarding your choice on expanding this interface ?

### CODE REVIEW

**Comment by six7** [comment]
This is awesome! 🔥 Let's schedule something for next week? This week is going to be still busy and we're about to merge the `release-114` branch which contains quite a few changes (probably also affecting some parts of this PR), but happy to jump on a call to discuss this, I think this would greatly benefit everyone!

Can you [DM me on twitter](https://twitter.com/six7) to discuss details?

**Comment by andrewx82** [comment]
@LiamMartens  Would you mind giving some feedback on changing the interface ?

**Comment by andrewx82** [comment]
@LiamMartens  bump

**Comment by LiamMartens** [comment]
@andrewx82 sorry i havent had time yet, I'm going to dive in asap

**Comment by andrewx82** [comment]
Hi @LiamMartens  any update? I know Jan mentioned he would prefer to use the remote URL storage provider to be the base of the this storage system. In the case that we establish a protocol of the interaction between the remote and local with support for branches, etc should we also include a basic express server to demo how it would work?

**Comment by six7** [comment]
@LiamMartens pinging again, you mentioned you left some comments but I guess those are still in pending state?

**Review by LiamMartens — CHANGES_REQUESTED** [review]

**Comment by LiamMartens** [comment]
@andrewx82 @six7 sorry I forgot to submit the comments...

File: src/constants/StorageProviderType.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
 export enum StorageProviderType {
   LOCAL = 'local',
+  GENERIC_VERSIONED_STORAGE = 'genericVersionedStorage',
   JSONBIN = 'jsonbin',
```
[the remark this answers is not in the document]
Comment (six7) [reply src/constants/StorageProviderType.ts:4]: 👍 yeah I'd favor a more generic one, now that we have `$metadata` there's no need to treat it special?

[the remark this answers is not in the document]
Comment (andrewx82) [reply src/constants/StorageProviderType.ts:4]: Regarding whether its writable or not , would we support both POST and PUT ? Would it follow the same setup as jsonbin with an initial POST to either create or check for existence and then  PUT afterwards ?

**Comment by andrewx82** [comment]
Thank you for the feedback @LiamMartens . Can I confirm with you that I should make the following changes to make the MR acceptable for you ?

- Add the utility function for `saveLastSyncedState`
- Eventlike support for the onChange event and the corresponding change for cloning. In this scenario I think it would be preferable that I pass through the entire array as you've suggested for now .

and then I just want to double check if you think the flow of POST to establish the endpoint and then PUTs and GETs afterwards for writes and reads respectively is fine as an interface.

I'll obviously resolve the merge conflicts as well.

**Comment by andrewx82** [comment]
Apologies for the delay, I have had to deal with some other work which blocked this. I will resolve all the above asap

**Pull request closed without merging**